usr/lib/soft_stdll/soft_specific.c

```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+                                          TEMPLATE *publ_tmpl,
+                                          TEMPLATE *priv_tmpl)
+{
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    CK_BBOOL rc;
+    CK_ATTRIBUTE *prime_attr = NULL;
+    CK_ATTRIBUTE *base_attr = NULL;
+    CK_ATTRIBUTE *temp_attr = NULL;
+    CK_ATTRIBUTE *value_bits_attr = NULL;
+    CK_BYTE *temp_byte;
+    CK_ULONG temp_bn_len;
+    DH *dh;
+    BIGNUM *bn_p;
+    BIGNUM *bn_g;
+    const BIGNUM *temp_bn;
+
+    UNUSED(tokdata);
+
+    rc = template_attribute_find(publ_tmpl, CKA_PRIME, &prime_attr);
+    rc &= template_attribute_find(publ_tmpl, CKA_BASE, &base_attr);
+
+    if (rc == FALSE) {
+        TRACE_ERROR("Could not find CKA_PRIME or CKA_BASE for the key\n");
+        return CKR_FUNCTION_FAILED;
+    }
+
+    if ((prime_attr->ulValueLen > 256) || (prime_attr->ulValueLen < 64)) {
+        TRACE_ERROR("CKA_PRIME attribute value is invalid.\n");
+        return CKR_ATTRIBUTE_VALUE_INVALID;
+    }
+
+    dh = DH_new();
+    if (dh == NULL) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return CKR_FUNCTION_FAILED;
+    }
+    // Create and init BIGNUM structs to stick in the DH struct
+    bn_p = BN_new();
+    bn_g = BN_new();
+    if (bn_g == NULL || bn_p == NULL) {
+        if (bn_g)
+            BN_free(bn_g);
+        if (bn_p)
+            BN_free(bn_p);
+        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+        return CKR_HOST_MEMORY;
+    }
+    // Convert from strings to BIGNUMs and stick them in the DH struct
+    BN_bin2bn((unsigned char *) prime_attr->pValue, prime_attr->ulValueLen,
+              bn_p);
+    BN_bin2bn((unsigned char *) base_attr->pValue, base_attr->ulValueLen, bn_g);
+    dh->p = bn_p;
+    dh->g = bn_g;
+
+    // Generate the DH Key
+    if (!DH_generate_key(dh)) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return CKR_FUNCTION_FAILED;
+    }
+    // Extract the public and private key components from the DH struct,
+    // and insert them in the publ_tmpl and priv_tmpl
+
+    //
+    // pub_key
+    //
+    //temp_bn = BN_new();
+    temp_bn = dh->pub_key;
+    temp_bn_len = BN_num_bytes(temp_bn);
+    temp_byte = malloc(temp_bn_len);
+    temp_bn_len = BN_bn2bin(temp_bn, temp_byte);
+    // in bytes
+    rc = build_attribute(CKA_VALUE, temp_byte, temp_bn_len, &temp_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        return CKR_FUNCTION_FAILED;
+    }
+    template_update_attribute(publ_tmpl, temp_attr);
+    free(temp_byte);
+
+    //
+    // priv_key
+    //
+    //temp_bn = BN_new();
+    temp_bn = dh->priv_key;
+    temp_bn_len = BN_num_bytes(temp_bn);
+    temp_byte = malloc(temp_bn_len);
+    temp_bn_len = BN_bn2bin(temp_bn, temp_byte);
+    // in bytes
+    rc = build_attribute(CKA_VALUE, temp_byte, temp_bn_len, &temp_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
```

Comment:
Shouldn't we free dh, bn_p, bn_g, and temp_byte here?

Reply:
done